### PR TITLE
Warn unsupported attributes

### PR DESCRIPTION
**Pull request by Happypig375**

### CODE REVIEW

**Review by dsyme — APPROVED** [review]

**Comment by dsyme** [comment]
Need to fix up the test and get it green, but this is great, thank you.

File: src/fsharp/PostInferenceChecks.fs [hunk 1]
```diff
@@ -1639,7 +1639,12 @@ and CheckDecisionTreeTest cenv env m discrim =
     | DecisionTreeTest.ActivePatternCase (exp, _, _, _, _, _) -> CheckExprNoByrefs cenv env exp
     | DecisionTreeTest.Error _ -> ()
 
-and CheckAttrib cenv env (Attrib(_, _, args, props, _, _, _)) = 
+and CheckAttrib cenv env (Attrib(tcref, _, args, props, _, _, m)) =
+    if [
```
Comment (dsyme) [inline src/fsharp/PostInferenceChecks.fs:1643]: This creates a list each time the function is called. It's no big deal for a list of length 2, but you can pre-create the list in TcGlobals if you like.

**Comment by KevinRansom** [comment]
@dsyme, I'm wondering if this should be an error, because if we should ever choose to implement the functionality declared by these attributes then it would be a breaking change.

**Comment by auduchinok** [comment]
@KevinRansom Making it an error where it'd previously compile fine would be a breaking change itself.

File: tests/FSharp.Compiler.ComponentTests/ErrorMessages/UnsupportedAttributes.fs [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+let [<ModuleInitializer>] g () = ()
+type C() =
+    member _.F (w, [<System.Runtime.CompilerServices.CallerArgumentExpression "w">] x : string) = ()
+    [<System.Runtime.CompilerServices.ModuleInitializer>]
+    member _.G() = ()
+        """
+        |> FSharp
+        |> typecheck
+        |> shouldFail
+        |> withResults [
```
Comment (vzarytovskii) [inline tests/FSharp.Compiler.ComponentTests/ErrorMessages/UnsupportedAttributes.fs:25]: Just a note - a less verbose `withDiagnostics` can be used here, for example, like in the `ErrorMessages` suite:
https://github.com/dotnet/fsharp/blob/dbf9a625d3188184ecb787a536ddb85a4ea7a587/tests/FSharp.Compiler.ComponentTests/ErrorMessages/ClassesTests.fs#L68-L71

**Review by vzarytovskii — APPROVED** [review]

File: tests/FSharp.Compiler.ComponentTests/ErrorMessages/UnsupportedAttributes.fs (hunk 2 above)
Comment (Happypig375) [reply tests/FSharp.Compiler.ComponentTests/ErrorMessages/UnsupportedAttributes.fs:25]: @vzarytovskii How can I generate them? For `withResults` I can just copy from the test output.

Comment (vzarytovskii) [reply tests/FSharp.Compiler.ComponentTests/ErrorMessages/UnsupportedAttributes.fs:25]: > @vzarytovskii How can I generate them? For `withResults` I can just copy from the test output.

I'm afraid, you'll have to write them manually :(
Let me see if I can serialise them into something which you can just copy-paste into `withDiagnostics`.

**Pull request merged**